### ISSUE TITLE
Can't reproduce results

### ISSUE THREAD

**Issue by AhmadObeid** [Yanfeng-Zhou/XNet#29]
Thank you for your nice code, it was very easy to run it. 
I experimented with the GlaS dataset used in your paper, but I can't seem to get the same results with XNet or any of the other methods you implement. 
You mention training for 200 epochs in the paper, but it seems the model still underfits the data at that stage. 
Indeed training for more epochs increases the performance on the validation set. 
I trained for 2000 epochs, but the results on the testing set is still much less than what's in the paper. 
I got JC 0.3953 and DC 0.5666 on the test set after 2000 epochs of training. 
I didn't change anything in your code, expect for the Dataset class minimally to control the labeled/unlabeled sets. 
Below are more details:

```
class dataset_iitnn(Dataset):
    def __init__(self, data_dir, input1, input2, augmentation1, normalize_1, normalize_2, sup=True,
                 num_images=None,split="train",lab_perc=0.2, **kwargs):
        super(dataset_iitnn, self).__init__()

        img_paths_1 = []
        img_paths_2 = []
        mask_paths = []

        image_dir_1 = os.path.join(data_dir,'wavelets',input1,split) 
        image_dir_2 = os.path.join(data_dir,'wavelets',input2,split) 
        
        if sup:
            mask_files = glob.glob(os.path.join(data_dir,split,'*','rgb/*_anno.bmp'))
            
        for image in os.listdir(image_dir_1):
            if image in ['zz.png','zz3.png']: continue  
            image_path_1 = os.path.join(image_dir_1, image)
            img_paths_1.append(image_path_1)

            image_path_2 = os.path.join(image_dir_2, image)
            img_paths_2.append(image_path_2)
            if sup:
                image_name = image.split('.')[0]
                label_tmp = [f.split('/')[-3] for f in mask_files if image_name+'_anno' in f][0]
                mask_dir = os.path.join(data_dir,split,label_tmp,'rgb')
                mask_path = os.path.join(mask_dir, image.replace('.bmp','_anno.bmp'))
                mask_paths.append(mask_path)
        assert len(img_paths_1) == len(img_paths_2)
        if sup:
            assert len(img_paths_1) == len(mask_paths)
            if split == "train":
                img_paths_1 = img_paths_1[:int(lab_perc*(len(img_paths_1)))]
                img_paths_2 = img_paths_2[:int(lab_perc*(len(img_paths_2)))]
                mask_paths = mask_paths[:int(lab_perc*(len(mask_paths)))]
        else:
            if split == "train":
                img_paths_1 = img_paths_1[int(lab_perc*(len(img_paths_1))):]
                img_paths_2 = img_paths_2[int(lab_perc*(len(img_paths_2))):]
                mask_paths = mask_paths[int(lab_perc*(len(mask_paths))):]

        self.img_paths_1 = img_paths_1
        self.img_paths_2 = img_paths_2
        self.mask_paths = mask_paths
        self.augmentation_1 = augmentation1
        self.normalize_1 = normalize_1
        self.normalize_2 = normalize_2
        self.sup = sup
        self.kwargs = kwargs

    def __getitem__(self, index):

        img_path_1 = self.img_paths_1[index]
        img_1 = Image.open(img_path_1)
        img_1 = np.array(img_1)

        img_path_2 = self.img_paths_2[index]
        img_2 = Image.open(img_path_2)
        img_2 = np.array(img_2)
        if self.sup:
            mask_path = self.mask_paths[index]
            mask = Image.open(mask_path)
            mask = np.array(mask)

            augment_1 = self.augmentation_1(image=img_1, image2=img_2, mask=mask)
            img_1 = augment_1['image']
            img_2 = augment_1['image2']
            mask = augment_1['mask']

            normalize_1 = self.normalize_1(image=img_1, mask=mask)
            img_1 = normalize_1['image']
            mask = normalize_1['mask']
            mask = mask.long()

            normalize_2 = self.normalize_2(image=img_2)
            img_2 = normalize_2['image']
            
            sampel = {'image': img_1, 'image_2': img_2, 'mask': mask, 'ID': os.path.split(mask_path)[1]}

        else:
            augment_1 = self.augmentation_1(image=img_1, image2=img_2)
            img_1 = augment_1['image']
            img_2 = augment_1['image2']

            normalize_1 = self.normalize_1(image=img_1)
            img_1 = normalize_1['image']

            normalize_2 = self.normalize_2(image=img_2)
            img_2 = normalize_2['image']

            sampel = {'image': img_1, 'image_2': img_2, 'ID': os.path.split(img_path_1)[1]}

        return sampel

    def __len__(self):
        return len(self.img_paths_1)
```

Inside the main python script, I always call the dataset class with num_images=None, and I handle the splitting into labeled and unlabeled as shown above. 

Here is the result of running test_xnet.py after 2000 epochs of trianing:

\===================================
| Test Thr: 0.8800                |
| Test  Jc: 0.3953                 |
| Test  Dc: 0.5666                |
\-----------------------------------
| Testing Completed In 0h 0mins 2s |
\===================================

**Comment by Yanfeng-Zhou** [comment]
Thank you for your recognition of our work. 

In the modification of dataloader, you seem to have not ensured that the labeled dataset and the unlabeled dataset have the same length?

**Comment by AhmadObeid** [comment]
Thank you for you response, @Yanfeng-Zhou. 
You are absolutely right it seems I somehow missed the part you mentioned in your response. 
Retrying training while making sure they have the same length indeed increased the performance; yet, it is still below what is in the paper. 
For context, I use the full GlaS train subset for training, and testB for validation, and testA for testing. Could this be the source of disparity between our results?

Here is what I got after 2000 epochs of training with 10% labels:

\===================================
| Test Thr: 0.8800                |
| Test  Jc: 0.5040                |
| Test  Dc: 0.6702                |
\-----------------------------------
| Testing Completed In 0h 0mins 3s |
\===================================
And here for 20% labels:
\ ===================================
| Test Thr: 0.8800                |
| Test  Jc: 0.4869                |
| Test  Dc: 0.6549                |
\-----------------------------------
| Testing Completed In 0h 0mins 3s |
\===================================

NOTE:
The performance of the other methods all decreased after equating the lengths. Is this normal behavior?
Thank you

**Comment by Yanfeng-Zhou** [comment]
The current results show that the model has not been trained successfully. For the GlaS dataset, have you ensured that the foreground in the mask is 1 and the background is 0?

Do you encounter the same error on other datasets?

**Comment by AhmadObeid** [comment]
Yes, I did ensure the foreground is 1 and background is 0. 
Trying on a different dataset (RINGS), and with 20% annotation, all methods achieve good results except XNet. 
What is weird is that at the end of the training, it confirms that it found the best threshold and results to be:

\| Training Completed In 11h 17mins 9s 
\---------------------------------------------------------------
| Best  Result: Result1                                       |
| Best Val Thr: 0.2800                                        |
| Best Val  Jc: 0.7142                                        |
| Best Val  Dc: 0.8333                                        |

However, when using test_xnet.py, I get a bad result and the threshold goes to zero. 
I inspected the predictions and they mostly lie near zero which explains the threshold problem, but this may mean that either the model is not saved properly during training, or it is not being loaded correctly. 
I even tried "testing" on the training and validation sets, expecting to get at least something near the results above, but still got very bad results.
Note that I wrote other test scripts (test_urpc.py, test_ct.py, test_uamt.py...etc) and got good results on RINGS.

**Comment by Yanfeng-Zhou** [comment]
I see. There is a BUG in the current XNet weight storage. Please do not use the weight corresponding to the best result, but use the weight with performance!

**Comment by AhmadObeid** [comment]
Thank you again @Yanfeng-Zhou, it seems you may be right; it is probably a saving/loading issue; however, it doesn't look like it is simply best vs the one with performance. 
Somehow some methods run into this problem and others don't. I will confirm later by retraining those who currently still have the same problem. 
For most methods now the results are sensible, although slightly different from your paper. For reference, here is a table of the results I am currently observing:

![Image](https://github.com/user-attachments/assets/21b309ee-91ca-4d87-a321-ed7de6fe40d4)

**Comment by Yanfeng-Zhou** [comment]
Thank you for your recognition of our work. For your current training results, CCT, URPC, CT, and XNet do not seem to have reached a good state. You can try to refine the hyperparameters to make the results better. 

For GlaS, if you use 85 images for training (17+68) and 80 images for testing, you can get a Jaccard (JI) of about 0.8.
